content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,388 @@
+    filterText: ""
+  };
+
+  handleChange = event => {
+    this.setState({ filterText: event.target.value });
+  };
+
+  render() {
+    // The render method on this PureComponent is not called unless either
+    // props.list or state.filterText have referentially different values.
+    const filteredList = this.props.list.filter(
+      item => item.text.includes(this.state.filterText)
+    )
+
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+}
+```
+
+The above approach is much cleaner and simpler than the derived state version, but it has some limitations. Filtering may be slow for large lists, and `PureComponent` won't prevent re-renders if props other than `list` change. In that case, we could add a memoization helper to avoid unnecessarily re-filtering our list:
+
+```js
+import memoize from "memoize-one";
+
+class Example extends Component {
+  // State only needs to hold the current filter text value:
+  state = {
+    filterText: ""
+  };
+
+  // Re-run the filter whenever the list array or filter text change:
+  filter = memoize(
+    (list, filterText) => list.filter(
+      item => item.text.includes(filterText)
+    )
+  );
+
+  handleChange = event => {
+    this.setState({ filterText: event.target.value });
+  };
+
+  render() {
+    const filteredList = this.filter(this.props.list, this.state.filterText);
+
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+}
+```
+
+This is much simpler and performs just as well as the previous version!
+
+There are a couple of constraints to consider when using memoization though:
+
+1. In most cases, you'll want to **attach the memoized function to a component instance**. This prevents multiple instances of a component from resetting each other's memoized keys.
+1. Typically you'll want to use a memoization helper with a **limited cache size** in order to prevent memory from "leaking" over time. (In the example above, we used `memoize-one` because it only caches the most recent argument and result.)
+1. Memoization will not work properly if `props.items` is recreated each time the parent component renders. Hopefully, this should not be the case for large arrays. If it were, you'd want to address that problem first! (This limitation also applies to the `getDerivedStateFromProps` version above.)
+
+## In closing
+
+The examples above are intentionally simplified in order to highlight specific coding patterns. In real world applications, components often contain a mix of controlled and uncontrolled behaviors. This is okay! Just be careful to ensure that each behavior has a clear source of truth in order to avoid the anti-patterns mentioned above.
```

Comment:
Maybe we could add a small blurb here saying that in general `getDerivedStateFromProps` should be used sparely, and that its verbosity exposes the complexity of what you're trying to do.